pkg/cac/cac.go

```diff
@@ -8,85 +8,87 @@ import (
 	"bytes"
 	"encoding/binary"
 	"errors"
+	"fmt"
 
 	"github.com/ethersphere/bee/pkg/bmtpool"
 	"github.com/ethersphere/bee/pkg/swarm"
 )
 
 var (
-	errTooShortChunkData = errors.New("short chunk data")
-	errTooLargeChunkData = errors.New("data too large")
+	ErrChunkDataShort = errors.New("chunk data length is not sufficient")
+	ErrChunkDataLarge = errors.New("chunk data exceeds maximum allowed length")
 )
 
 // New creates a new content address chunk by initializing a span and appending the data to it.
 func New(data []byte) (swarm.Chunk, error) {
 	dataLength := len(data)
-	if dataLength > swarm.ChunkSize {
-		return nil, errTooLargeChunkData
-	}
 
-	if dataLength == 0 {
-		return nil, errTooShortChunkData
+	if err := validateDataLength(dataLength); err != nil {
+		return nil, fmt.Errorf("invalid CAC data length (%d): %w", dataLength, err)
 	}
 
 	span := make([]byte, swarm.SpanSize)
 	binary.LittleEndian.PutUint64(span, uint64(dataLength))
+
 	return newWithSpan(data, span)
 }
 
 // NewWithDataSpan creates a new chunk assuming that the span precedes the actual data.
 func NewWithDataSpan(data []byte) (swarm.Chunk, error) {
 	dataLength := len(data)
-	if dataLength > swarm.ChunkSize+swarm.SpanSize {
-		return nil, errTooLargeChunkData
-	}
 
-	if dataLength < swarm.SpanSize {
-		return nil, errTooShortChunkData
+	if err := validateDataLength(dataLength - swarm.SpanSize); err != nil {
+		return nil, fmt.Errorf("invalid CAC data length (%d): %w", dataLength, err)
 	}
+
 	return newWithSpan(data[swarm.SpanSize:], data[:swarm.SpanSize])
 }
 
+func validateDataLength(dataLength int) error {
+	if dataLength <= 0 {
```

Comment:
I don't know a lot about spans, but does the change from < swarm.SpanSize to <= 0 invalidate a CAC of a zero-length file?  swarm.SpanSize has already been removed in the call to this method.

Reply:
Currently, `cac` has varying validation for edge case when data == 0.

When `cac` is created with `New` function, error will be returned if `len(data) == 0`
https://github.com/ethersphere/bee/blob/41d6eefa62838830c934f912cd7e87e150ccd68f/pkg/cac/cac.go#L28

When `cac` is created with `NewWithDataSpan` function, it would be possible to create `cac` with zero length data (avoiding condition shown above) if `len(data) == swarm.SpanSize` https://github.com/ethersphere/bee/blob/41d6eefa62838830c934f912cd7e87e150ccd68f/pkg/cac/cac.go#L44

This PR unifies validation to and follows through decision to forbid case when `data == 0`. If this needs to be corrected (to allow this case), validation condition will be updated.

Reply:
I suspect this new decision forbidding data==0 will fail the upload of a zero length file.  Or zero length content with the /bytes endpoint, which is effectively the same thing.  I can forsee many data structures that may require a reference to zero-length content, which this change will no longer allow storing within the swarm.